### PR TITLE
Use implicit status of further arguments in measure/wf-like Program Fixpoint

### LINKED ISSUE #13697: Program Fixpoint breaks recursive calls with explicit arguments
#### Version

Coq 8.12.0

#### Operating system

Red Hat Enterprise Linux Workstation 7.4

#### Description of the problem

The following program fails for no reason.

```coq
From Coq Require micromega.Lia Lists.List Program.Program.

Set Implicit Arguments.
Set Strict Implicit. Set Strongly Strict Implicit.
Set Maximal Implicit Insertion.

Section Context.

Import Lia List Tactics.
Import ListNotations.

Obligation Tactic := idtac.

Program Fixpoint merge (A : Type) (f : A -> A -> bool) (l0 l1 : list A)
  {measure (length l0 + length l1)} : list A :=
  match l0, l1 with
  | [], _ => l1
  | _, [] => l0
  | n0 :: k0, n1 :: k1 => if f n0 n1 then
    n0 :: merge f k0 l1 else
    n1 :: @merge _ f l0 k1
  end.
Next Obligation. intros; subst; cbn in *; lia. Qed.
Next Obligation. intros; subst; cbn in *; lia. Qed.
Next Obligation. program_simplify. Qed.
Next Obligation. program_solve_wf. Defined.

End Context.

Section Context.

Import List NArith.
Import ListNotations N.

Open Scope N_scope.

Compute merge leb [7; 13; 42] [18; 69; 420].

End Context.
```

```
In environment
A : Type
f : A -> A -> bool
l0 : list A
l1 : list A
merge :
forall A0 : Type,
(A0 -> A0 -> bool) -> forall l2 l3 : list A0, length l2 + length l3 < length l0 + length l1 -> list A0
n0 : A
k0 : list A
n1 : A
k1 : list A
Heq_l0 : n0 :: k0 = l0
Heq_l1 : n1 :: k1 = l1
The term "merge A f l0 k1" has type "length l0 + length k1 < length l0 + length l1 -> list A"
while it is expected to have type "list A".
```

#### Notes

Changing the call `@merge _ f l0 k1`
into `@merge _ f l0 k1 _` or `merge (A := A) f l0 k1` avoids the problem,
because it allows inferring the hidden implicit argument.
Whether this is a bug or a feature is unclear.

### LINKED ISSUE #13812: Program Fixpoint with wf fixannot and function return types
There seems to be a problem in the proof term generated by Program Fixpoint when the wf fixannot is set and when the return type is a function type.

The problem occurs in the recursive call to foo in the following example:

```coq
Require Import Coq.Program.Wf.

Inductive Nat : Type :=
| O : Nat
| S : Nat -> Nat.

Inductive wfn : Nat -> Nat -> Prop :=
| wfn_suc : forall n, wfn n (S n)
| wfn_trans : forall x y z, wfn x y -> wfn y z -> wfn x z
.

Definition wfn_Sn (n : Nat) : Acc wfn (S n).
Admitted.

Lemma wfn_wf : well_founded wfn.
Admitted.

Program Fixpoint foo (n : Nat) {wf wfn n }: Nat -> Nat := fun x =>
  match n with
    O => x
  | S n => foo n _
  end.
Obligation 1. apply wfn_suc. Qed.
Obligation 3. unfold MR. apply wfn_wf. Qed.
```

`foo n _` should be of type `Nat -> Nat` and the program shouldn't typecheck as a term of type `Nat` is expected.
Instead, it seems the argument is treated as implicit and the program is accepted. 
When inspecting the produced proof obligation, one can see that the term `x` is passed to `foo`.
This problem does not occur when no fixannot is set.

My coq version is 8.13.

### LINKED ISSUE #14841: Program Fixpoint's `measure` conflicts with notations in `where` clause.
#### Description of the problem

Coq does not accept a `Program Fixpoint` with `measure` and notation in `where` clause.

```coq
From Coq Require Import Program.

Reserved Notation "'bar' x" (at level 60).

(* works *)
(*
Fixpoint foo (n : nat) :=
  match n with
  | 0 => 0
  | S n' => S (bar n')
  end
where "'bar' x" := (foo x).
 *)

(* works *)
(*
Program Fixpoint goo (n : nat) :=
  match n with
  | 0 => 0
  | S n' => S (bar n')
  end
where "'bar' x" := (goo x).
*)

(* Works *)
(*
Program Fixpoint goo (n : nat) {struct n} :=
  match n with
  | 0 => 0
  | S n' => S (bar n')
  end
where "'bar' x" := (goo x).
*)


(* Does not work. `Error: Unknown interpretation for notation "bar _".`
 *)
Program Fixpoint goo (n : nat) {measure n} :=
  match n with
  | 0 => 0
  | S n' => S (bar n')
  end
where "'bar' x" := (goo x).
```

#### Coq Version
tested on version 8.13.2 and current master branch (bdb72aab60f40172743de2bbfd111c22037652c3)

### PR DESCRIPTION
**Pull request by herbelin**
The purpose of the PR is to align the behavior of `measure`/`wf`-like `Program Fixpoint` on what is done for `Fixpoint` in general. This includes:
- taking into account the presence of `forall` and of implicit arguments in the conclusion of the fixpoint
- supporting the `where` clause for notations
- supporting the `Local` modifier / `local` attribute
- supporting the `clearbody` attribute

For instance, the PR makes that the following does not fail anymore:
```coq
Require Import Program.
Program Fixpoint zero (n : nat) {measure n} : nat -> nat :=
  match n with
    | 0 => fun _ => 0
    | S n' => zero n'
  end.

Program Fixpoint f n {B} (b:B) {measure n} : forall {A}, A -> A * B :=
  match n with
  | 0 => fun A a => (a, b)
  | S n => fun A a => f n b a
  end.
```

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

Fixes #14841 (regarding the `where` clause)
Fixes #13812 (regarding further arguments)

Note: it does not fix #13697, which would require hacking recursive calls made with `@` (see #18878 instead).

### CODE REVIEW

**Comment by herbelin** [comment]
@coqbot run full ci

**Review by SkySkimmer — APPROVED** [review]
Changelog?

**Comment by herbelin** [comment]
> Changelog?

I tried to concoct something.

File: doc/changelog/02-specification-language/18834-master+use-extra-impargs-program-fixpoint-wf.rst [hunk 1]
```diff
@@ -0,0 +1,6 @@
+- **Added:**
+  :cmd:`Program Fixpoint` with `measure` or `wf` now supports the
```
Comment (SkySkimmer) [inline doc/changelog/02-specification-language/18834-master+use-extra-impargs-program-fixpoint-wf.rst:2]: Looks like the refman doesn't treat Program Fixpoint as its own command.
```suggestion
  :ref:`program_fixpoint` with `measure` or `wf` now supports the
```

Comment (herbelin) [reply doc/changelog/02-specification-language/18834-master+use-extra-impargs-program-fixpoint-wf.rst]: Thanks. I removed the `:cmd:` and added the `:ref:`.

File: doc/changelog/02-specification-language/18834-master+use-extra-impargs-program-fixpoint-wf.rst [hunk 2]
```diff
@@ -0,0 +1,9 @@
+n- **Added:**
```
Comment (SkySkimmer) [inline doc/changelog/02-specification-language/18834-master+use-extra-impargs-program-fixpoint-wf.rst:1]:
```suggestion
- **Added:**
```

**Pull request merged**